cilium/CFP-14339-cilium-mtu-model.md

```diff
@@ -0,0 +1,292 @@
+widely, such as allowing 1500 byte ethernet payloads via one network while
+supporting as much as 9000 byte payloads via another network. Misconfiguration
+of the MTU across different nodes can cause connectivity disruption,
+particularly for UDP packets.
+
+In some user environments, the Cilium 1.9 MTU calculation behaviour is
+inconsistent as it is determined by either:
+
+* `--mtu` flag (administrator override)
+* `InitDefaultPrefix()` configures the node IP address
+   * `->firstGlobalV4Addr()`
+      * List all IPs from interfaces, except for interfaces that are down and
+        docker interfaces
+      * Pick the first one in the sorted list
+
+
+We also currently expose only MTU and not MRU. Rather than considering transmit
+and receive separately to determine the maximum sizes that can be viably sent
+vs. received depending on the attached networks, we select a single MTU and
+derive the MRU based on the configured/detected MTU and other configuration
+such as ipsec, tunnel. This means that if the MTU we detected is incorrect,
+then we may configure the MRU in a way that may cause traffic drops on receive
+in cases where the packets could otherwise have been successfully received.
+
+### Goals
+
+1. Consistently calculate MTU and MRU (Maximum Receive Unit) sizes on each node.
+2. Guarantee that packets of any reasonable size may be successfully received,
+   regardless of the attached networks.
+3. Provide visibility into the MTU and MRU size calculation and status on each
+   node.
+4. In cloud environments with East-West MTU 9K + North-South MTU 1.5K,
+   automatically configure in a way that allows pods to utilize 9K MTU.[^b][^c]
+5. Nice-to-haves:
+   1. Minimize the likelihood of relying on IP fragmentation for connectivity
+   2. Provide best-effort optimization of MTU/MRU configuration with ability
+      for administrators to override with simple configuration.
+
+### Non-Goals
+
+* Assume that Kubernetes view of the network is the correct view, ignoring
+  other networks
+
+## Proposal
+
+### Overview
+
+Regardless of whatever configuration you have, by default Cilium should be able
+to successfully establish connectivity from local endpoints and the local stack
+to arbitrary endpoints on any connected network. This might not provide optimal
+performance depending on the networks available to the node, but in that case
+we should provide visibility and configuration options so that administrators
+can understand how Cilium is determining MTU/MRU, and configure Cilium in a way
+that satisfies their use case.
+
+Use the available network devices on the node to calculate a maximum
+transmission unit (MTU) size and maximum receive unit (MRU) size separately in
+such a way that the lowest attached unit size will be the default MTU
+configured in routes[^d] for Cilium pods and the highest attached unit size
+will be the default MRU configured on devices for Cilium pods.
```

Comment:
```suggestion
```

This CFP was originally intending to simply make MTU detection + application logic more consistent and observable, then provide the administrator the tools to configure more advanced MTU handling behaviour. However, based on feedback and discussion, there seems to be desire to go further and automatically ensure optimal MTU configuration for common use cases.